Netezos/Forging/Local/LocalForge.Unforgers.Operations.cs

```diff
@@ -0,0 +1,251 @@
+﻿using Netezos.Encoding;
+using Netezos.Forging.Models;
+using Netezos.IO;
+using System;
+using System.Linq;
+
+namespace Netezos.Forging
+{
+    public partial class LocalForge
+    {
+        static OperationContent UnforgeOperation(MichelineReader reader)
+        {
+            int operation = (int)reader.ReadUBigInt();
+
+            switch ((OperationTag)operation)
+            {
+                case OperationTag.Endorsement:
+                    return UnforgeEndorsement(reader);
+                case OperationTag.Ballot:
+                    return UnforgeBallot(reader);
+                case OperationTag.Proposals:
+                    return UnforgeProposals(reader);
+                case OperationTag.Activation:
+                    return UnforgeActivation(reader);
+                case OperationTag.DoubleBaking:
+                    return UnforgeDoubleBaking(reader);
+                case OperationTag.DoubleEndorsement:
+                    return UnforgeDoubleEndorsement(reader);
+                case OperationTag.SeedNonceRevelation:
+                    return UnforgeSeedNonceRevelaion(reader);
+                case OperationTag.Delegation:
+                    return UnforgeDelegation(reader);
+                case OperationTag.Origination:
+                    return UnforgeOrigination(reader);
+                case OperationTag.Transaction:
+                    return UnforgeTransaction(reader);
+                case OperationTag.Reveal:
+                    return UnforgeReveal(reader);
+                default:
+                    throw new ArgumentException($"Invalid operation: {operation}");
+            }
+        }
+
+        static EndorsementContent UnforgeEndorsement(MichelineReader reader)
+        {
+            return new EndorsementContent
+            {
+                Level = reader.ReadInt32()
+            };
+        }
+
+        static BallotContent UnforgeBallot(MichelineReader reader)
+        {
+            return new BallotContent
+            {
+                Source = reader.ReadTzAddress(),
+                Period = reader.ReadInt32(),
+                Proposal = reader.ReadBase58(Lengths.P.Decoded, Prefix.P),
+                Ballot = (Ballot)reader.ReadByte()
+            };
+        }
+
+        static ProposalsContent UnforgeProposals(MichelineReader reader)
+        {
+            return new ProposalsContent
+            {
+                Source = reader.ReadTzAddress(),
+                Period = reader.ReadInt32(),
+                Proposals = reader.ReadEnumerable(r => r.ReadBase58(Lengths.P.Decoded, Prefix.P)).ToList()
+            };
+        }
+
+        static ActivationContent UnforgeActivation(MichelineReader reader)
+        {
+            return new ActivationContent
+            {
+                Address = reader.ReadTz1Address(),
+                Secret = Hex.Convert(reader.ReadBytesToEnd())
```

Comment:
Reading to the end can cause problems. Better to read the exact number of bytes to be sure the result is correct.
````cs
Secret = Hex.Convert(reader.ReadBytes(20))
````